docs/async-await.md

```diff
@@ -0,0 +1,232 @@
+  
+    public struct Iterator: AsyncIteratorProtocol {
+      public typealias Element = ByteBuffer
+    
+      public func next() async throws -> ByteBuffer?
+    }
+  
+    public func makeAsyncIterator() -> Iterator
+  }
+}
+```
+
+Note: The user must consume the `Body` stream or drop the `HTTPClientResponse`, to ensure that the 
+internal HTTPClient connection can move forward. Dropping the `HTTPClientResponse` would lead to a 
+request cancellation which in turn would lead to a close of an exisiting HTTP/1.1 connection.
+
+At a later point we could add trailers to the `HTTPClientResponse` as effectful properties:
+
+```swift
+    public var trailers: HTTPHeaders? { async throws }
+```
+
+However we will need to make sure that the user has consumed the body stream completely before, calling the trailers, because otherwise we might run into a situation from which we can not progress forward:
+
+```swift
+do {
+  var request = HTTPClientRequest(url: "https://swift.org/")
+  let response = try await httpClient.execute(request, deadline: .now() + .seconds(3))
+  
+  var trailers = try await response.trailers // can not move forward since body must be consumed before.
+} catch {
+  print(error)
+}
+```
+
+In such a case we can either throw an error or crash.
+
+### New invocation
+
+The new way to invoke a request shall look like this:
+
+```swift
+extension HTTPClient {
+  func execute(_ request: HTTPClientRequest, deadline: NIODeadline) async throws -> HTTPClientResponse
+}
+```
+
+Simple usage example:
+
+```swift
+var request = HTTPClientRequest(url: "https://swift.org")
+request.method = .POST
+request.headers = [
+  "content-type": "text/plain; charset=UTF-8"
+  "x-my-fancy-header": "super-awesome"
+]
+request.body = .bytes("Hello world!".utf8)
+
+var response = try await client.execute(request, deadline: .now() + .seconds(5))
+
+switch response.status {
+case .ok:
+  let body = try await response.body.collect(maxBytes: 1024 * 1024)
+default:
+  throw MyUnexpectedHTTPStatusError
+}
+```
+
+Stream upload and download example using the new `FileHandle` api:
+
+```swift
+let readHandle = FileHandle(forReadingFrom: URL(string: "file:///tmp/readfile")!))
+let writeHandle = FileHandle(forWritingTo: URL(string: "file:///tmp/writefile")!))
+
+var request = HTTPClientRequest(url: "https://swift.org/echo")
+request.method = .POST
+request.headers = [
+  "content-type": "text/plain; charset=UTF-8"
+  "x-my-fancy-header": "super-awesome"
+]
+request.body = .stream(readHandle.bytes)
+
+var response = try await client.execute(request, deadline: .now() + .seconds(5))
+
+switch response.status {
+case .ok:
+  Task {
+    var streamIterator = response.body.makeAsyncIterator()
+    writeHandle.writeabilityHandler = { handle in
+      switch try await streamIterator.next() {
+      case .some(let buffer):
+        handle.write(buffer.readData(buffer.readableBytes)!)
+      case .none:
+        handle.close()
+      }
```

Comment:
In general, yes because `response.body` implements `AsyncSequence`. However, in this specific example we want to support proper backpressure meaning that we don't want to read more from the body if the writeHandler is not ready to write more data. We therefore wait until we read the next buffer from `response.body` until `writeHandle.writeabilityHandler` is called.